python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1849,13 +1859,25 @@ def transfer_from(self, source_datastore: Datastore, refs: Iterable[DatasetRef],
         # Need to map these missing IDs to a DatasetRef so we can guess
         # the details.
         if missing_ids:
-            log.info("Number of expected datasets missing from source datastore records: %d",
-                     len(missing_ids))
+            log.info("Number of expected datasets missing from source datastore records: %d out of %d",
+                     len(missing_ids), len(requested_ids))
             id_to_ref = {ref.id: ref for ref in refs if ref.id in missing_ids}
 
             for missing in missing_ids:
-                expected = self._get_expected_dataset_locations_info(id_to_ref[missing])
-                source_records[missing].extend(info for _, info in expected)
+                # Ask the source datastore where the missing artifacts
+                # should be.  An execution butler might not know about the
+                # artifacts even if they are there.
+                expected = source_datastore._get_expected_dataset_locations_info(id_to_ref[missing])
+
+                # Not all components can be guaranteed to exist so this
+                # list has to filter those by checking to see if the
+                # artifact is really there.
+                records = [info for location, info in expected if location.uri.exists()]
+                if records:
+                    source_records[missing].extend(records)
+                else:
+                    log.warning("Asked to transfer dataset %s but no file artifacts exist for it.",
+                                missing)
```

Comment:
Maybe also print `id_to_ref[missing]` for more complete diagnostics if it can help?

Reply:
I was pondering that. The UUID is enough to find it but more information won't hurt.